cps/kni.c

```diff
@@ -668,6 +668,231 @@ rd_modroute(const struct nlmsghdr *req, struct route_update *update)
+
+static void
+rd_getroute_ipv4(struct cps_config *cps_conf, struct gk_lpm *ltbl,
+	struct mnl_nlmsg_batch *batch, const struct nlmsghdr *req, int family)
+{
+	struct rte_lpm_iterator_state state;
+	const struct rte_lpm_rule *re4;
+	int index;
+
+	int ret = rte_lpm_iterator_state_init(ltbl->lpm, 0, 0, &state);
+	if (ret < 0) {
+		CPS_LOG(ERR, "Failed to initialize the IPv4 LPM rule iterator state in %s\n",
+			__func__);
+		return;
+	}
+
+	index = rte_lpm_rule_iterate(&state, &re4);
+	while (index >= 0) {
+		struct gk_fib *fib = &ltbl->fib_tbl[re4->next_hop];
+		struct ipaddr *gw_addr;
+		struct nlmsghdr *reply;
+
+		if (fib->action != GK_FWD_GATEWAY_FRONT_NET &&
+				fib->action != GK_FWD_GATEWAY_BACK_NET &&
+				fib->action != GK_FWD_GRANTOR) {
+			index = rte_lpm_rule_iterate(&state, &re4);
+			continue;
+		}
+
+		reply = mnl_nlmsg_put_header(mnl_nlmsg_batch_current(batch));
+
+		rd_fill_getroute_reply(cps_conf, reply, fib,
+			family, req->nlmsg_seq, state.depth, &gw_addr);
+
+		/* Add address and gateway. */
+		mnl_attr_put_u32(reply, RTA_DST, htonl(re4->ip));
+		mnl_attr_put_u32(reply, RTA_GATEWAY, gw_addr->ip.v4.s_addr);
+
+		if (!mnl_nlmsg_batch_next(batch)) {
+			rd_send_batch(cps_conf, batch,
+				req->nlmsg_seq, req->nlmsg_pid);
+			mnl_nlmsg_batch_reset(batch);
+		}
+
+		index = rte_lpm_rule_iterate(&state, &re4);
+	}
+}
+
+static void
+rd_getroute_ipv6(struct cps_config *cps_conf, struct gk_lpm *ltbl,
+	struct mnl_nlmsg_batch *batch, const struct nlmsghdr *req, int family)
+{
+	struct rte_lpm6_iterator_state state6;
+	struct rte_lpm6_rule re6;
+	int index;
+
+	int ret = rte_lpm6_iterator_state_init(ltbl->lpm6, 0, 0, &state6);
+	if (ret < 0) {
+		CPS_LOG(ERR, "Failed to initialize the IPv6 LPM rule iterator state in %s\n",
+			__func__);
+		return;
+	}
+
+	index = rte_lpm6_rule_iterate(&state6, &re6);
+	while (index >= 0) {
+		struct gk_fib *fib = &ltbl->fib_tbl6[re6.next_hop];
+		struct ipaddr *gw_addr;
+		struct nlmsghdr *reply;
+
+		if (fib->action != GK_FWD_GATEWAY_FRONT_NET &&
+				fib->action != GK_FWD_GATEWAY_BACK_NET &&
+				fib->action != GK_FWD_GRANTOR) {
+			index = rte_lpm6_rule_iterate(&state6, &re6);
+			continue;
+		}
+
+		reply = mnl_nlmsg_put_header(mnl_nlmsg_batch_current(batch));
+
+		rd_fill_getroute_reply(cps_conf, reply, fib,
+			family, req->nlmsg_seq, state6.depth, &gw_addr);
+
+		/* Add address and gateway. */
+		mnl_attr_put(reply, RTA_DST,
+			sizeof(struct in6_addr), re6.ip);
+		mnl_attr_put(reply, RTA_GATEWAY,
+			sizeof(struct in6_addr), &gw_addr->ip.v6);
+
+		if (!mnl_nlmsg_batch_next(batch)) {
+			rd_send_batch(cps_conf, batch,
+				req->nlmsg_seq, req->nlmsg_pid);
+			mnl_nlmsg_batch_reset(batch);
+		}
+
+		index = rte_lpm6_rule_iterate(&state6, &re6);
+	}
+}
+
+static int
+rd_getroute(const struct nlmsghdr *req)
+{
+	/*
+	 * Buffer length set according to libmnl documentation:
+	 * the buffer that you have to use to store the batch must be
+	 * double of MNL_SOCKET_BUFFER_SIZE to ensure that the last
```

Comment:
Shouldn't we put the comments just right before the definition of the buffer, i.e., ``char buf[2 * MNL_SOCKET_BUFFER_SIZE];``